// eslint-etc/dist/cjs/from-fixture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromFixture = void 0;
function fromFixture(fixture, invalidTestCase = {}) {
    return {
        ...invalidTestCase,
        ...parseFixture(fixture),
    };
}
exports.fromFixture = fromFixture;
function parseFixture(fixture) {
    const errorRegExp = /^(\s*)(~+)\s*\[(\w+)\s*(.*?)\]\s*$/;
    const lines = [];
    const errors = [];
    fixture.split("\n").forEach((line) => {
        const match = line.match(errorRegExp);
        if (match) {
            const column = match[1].length + 1;
            const endColumn = column + match[2].length;
            const { length } = lines;
            errors.push({
                column,
                data: JSON.parse(match[4] || "{}"),
                endColumn,
                endLine: length,
                line: length,
                messageId: match[3],
            });
        }
        else {
            lines.push(line);
        }
    });
    return {
        code: lines.join("\n"),
        errors,
    };
}
