// eslint-plugin-rxjs/dist/rules/ban-operators.js
"use strict";
const common_tags_1 = require("common-tags");
const utils_1 = require("../utils");
const defaultOptions = [];
const rule = utils_1.ruleCreator({
    defaultOptions,
    meta: {
        docs: {
            category: "Best Practices",
            description: "Forbids the use of banned operators.",
            recommended: false,
        },
        fixable: undefined,
        messages: {
            forbidden: "RxJS operator is banned: {{name}}{{explanation}}.",
        },
        schema: [
            {
                type: "object",
                description: common_tags_1.stripIndent `
          An object containing keys that are names of operators
          and values that are either booleans or strings containing the explanation for the ban.`,
            },
        ],
        type: "problem",
    },
    name: "ban-operators",
    create: (context, unused) => {
        let bans = [];
        const [config] = context.options;
        if (!config) {
            return {};
        }
        Object.entries(config).forEach(([key, value]) => {
            if (value !== false) {
                bans.push({
                    explanation: typeof value === "string" ? value : "",
                    regExp: new RegExp(`^${key}$`),
                });
            }
        });
        function getFailure(name) {
            for (let b = 0, length = bans.length; b < length; ++b) {
                const ban = bans[b];
                if (ban.regExp.test(name)) {
                    const explanation = ban.explanation ? `: ${ban.explanation}` : "";
                    return {
                        messageId: "forbidden",
                        data: { name, explanation },
                    };
                }
            }
            return undefined;
        }
        return {
            "ImportDeclaration[source.value='rxjs/operators'] > ImportSpecifier": (node) => {
                const identifier = node.imported;
                const failure = getFailure(identifier.name);
                if (failure) {
                    context.report({
                        ...failure,
                        node: identifier,
                    });
                }
            },
        };
    },
});
module.exports = rule;
