// eslint-plugin-rxjs/dist/rules/no-finnish.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        docs: {
            category: "Stylistic Issues",
            description: "Forbids the use of Finnish notation.",
            recommended: false,
        },
        fixable: undefined,
        messages: {
            forbidden: "Finnish notation is forbidden.",
        },
        schema: [],
        type: "problem",
    },
    name: "no-finnish",
    create: (context) => {
        const { esTreeNodeToTSNodeMap } = eslint_etc_1.getParserServices(context);
        const { couldBeObservable, couldReturnObservable } = eslint_etc_1.getTypeServices(context);
        function checkNode(nameNode, typeNode) {
            if (couldBeObservable(typeNode || nameNode) ||
                couldReturnObservable(typeNode || nameNode)) {
                const tsNode = esTreeNodeToTSNodeMap.get(nameNode);
                if (/[$]+$/.test(tsNode.getText())) {
                    context.report({
                        loc: eslint_etc_1.getLoc(tsNode),
                        messageId: "forbidden",
                    });
                }
            }
        }
        return {
            "ArrayPattern > Identifier[name=/[$]+$/]": (node) => checkNode(node),
            "ArrowFunctionExpression > Identifier[name=/[$]+$/]": (node) => {
                const parent = eslint_etc_1.getParent(node);
                if (node !== parent.body) {
                    checkNode(node);
                }
            },
            "ClassProperty[key.name=/[$]+$/] > Identifier": (node) => checkNode(node, eslint_etc_1.getParent(node)),
            "FunctionDeclaration > Identifier[name=/[$]+$/]": (node) => {
                const parent = eslint_etc_1.getParent(node);
                if (node === parent.id) {
                    checkNode(node, parent);
                }
                else {
                    checkNode(node);
                }
            },
            "FunctionExpression > Identifier[name=/[$]+$/]": (node) => {
                const parent = eslint_etc_1.getParent(node);
                if (node === parent.id) {
                    checkNode(node, parent);
                }
                else {
                    checkNode(node);
                }
            },
            "MethodDefinition[key.name=/[$]+$/]": (node) => checkNode(node.key, node),
            "ObjectExpression > Property[computed=false][key.name=/[$]+$/]": (node) => checkNode(node.key),
            "ObjectPattern > Property[value.name=/[$]+$/]": (node) => checkNode(node.value),
            "TSCallSignatureDeclaration > Identifier[name=/[$]+$/]": (node) => checkNode(node),
            "TSConstructSignatureDeclaration > Identifier[name=/[$]+$/]": (node) => checkNode(node),
            "TSParameterProperty > Identifier[name=/[$]+$/]": (node) => checkNode(node),
            "TSPropertySignature > Identifier[name=/[$]+$/]": (node) => checkNode(node, eslint_etc_1.getParent(node)),
            "TSMethodSignature > Identifier[name=/[$]+$/]": (node) => {
                const parent = eslint_etc_1.getParent(node);
                if (node === parent.key) {
                    checkNode(node, parent);
                }
                else {
                    checkNode(node);
                }
            },
            "VariableDeclarator[id.name=/[$]+$/]": (node) => checkNode(node.id, node.init || node),
        };
    },
});
module.exports = rule;
