// eslint-plugin-rxjs/dist/rules/no-ignored-subscribe.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        docs: {
            category: "Best Practices",
            description: "Forbids the calling of `subscribe` without specifying arguments.",
            recommended: false,
        },
        fixable: undefined,
        messages: {
            forbidden: "Calling subscribe without arguments is forbidden.",
        },
        schema: [],
        type: "problem",
    },
    name: "no-ignored-subscribe",
    create: (context) => {
        const { couldBeObservable } = eslint_etc_1.getTypeServices(context);
        return {
            "CallExpression[arguments.length = 0][callee.property.name='subscribe']": (node) => {
                const callee = node.callee;
                if (couldBeObservable(callee.object)) {
                    context.report({
                        messageId: "forbidden",
                        node: callee.property,
                    });
                }
            },
        };
    },
});
module.exports = rule;
