// eslint-plugin-rxjs/dist/rules/no-ignored-error.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        docs: {
            category: "Best Practices",
            description: "Forbids the calling of `subscribe` without specifying an error handler.",
            recommended: false,
        },
        fixable: undefined,
        messages: {
            forbidden: "Calling subscribe without an error handler is forbidden.",
        },
        schema: [],
        type: "problem",
    },
    name: "no-ignored-error",
    create: (context) => {
        const { couldBeObservable, couldBeFunction } = eslint_etc_1.getTypeServices(context);
        return {
            "CallExpression[arguments.length > 0] > MemberExpression > Identifier[name='subscribe']": (node) => {
                const memberExpression = eslint_etc_1.getParent(node);
                const callExpression = eslint_etc_1.getParent(memberExpression);
                if (callExpression.arguments.length < 2 &&
                    couldBeObservable(memberExpression.object) &&
                    couldBeFunction(callExpression.arguments[0])) {
                    context.report({
                        messageId: "forbidden",
                        node,
                    });
                }
            },
        };
    },
});
module.exports = rule;
