// eslint-plugin-rxjs/dist/rules/no-ignored-takewhile-value.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        docs: {
            category: "Possible Errors",
            description: "Forbids ignoring the value within `takeWhile`.",
            recommended: "error",
        },
        fixable: undefined,
        messages: {
            forbidden: "Ignoring the value within takeWhile is forbidden.",
        },
        schema: [],
        type: "problem",
    },
    name: "no-ignored-takewhile-value",
    create: (context) => {
        function checkNode(expression) {
            let ignored = true;
            const scope = context.getScope();
            const [param] = expression.params;
            if (param && eslint_etc_1.isIdentifier(param)) {
                const variable = scope.variables.find(({ name }) => name === param.name);
                if (variable && variable.references.length > 0) {
                    ignored = false;
                }
            }
            if (ignored) {
                context.report({
                    messageId: "forbidden",
                    node: expression,
                });
            }
        }
        return {
            "CallExpression[callee.name='takeWhile'] > ArrowFunctionExpression": (node) => checkNode(node),
            "CallExpression[callee.name='takeWhile'] > FunctionExpression": (node) => checkNode(node),
        };
    },
});
module.exports = rule;
