// eslint-plugin-rxjs/dist/rules/no-ignored-observable.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        docs: {
            category: "Best Practices",
            description: "Forbids the ignoring of observables returned by functions.",
            recommended: false,
        },
        fixable: undefined,
        messages: {
            forbidden: "Ignoring a returned Observable is forbidden.",
        },
        schema: [],
        type: "problem",
    },
    name: "no-ignored-observable",
    create: (context) => {
        const { couldBeObservable } = eslint_etc_1.getTypeServices(context);
        return {
            "ExpressionStatement > CallExpression": (node) => {
                if (couldBeObservable(node)) {
                    context.report({
                        messageId: "forbidden",
                        node,
                    });
                }
            },
        };
    },
});
module.exports = rule;
