// eslint-plugin-rxjs/dist/rules/no-ignored-replay-buffer.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        docs: {
            category: "Best Practices",
            description: "Forbids using `ReplaySubject`, `publishReplay` or `shareReplay` without specifying the buffer size.",
            recommended: "error",
        },
        fixable: undefined,
        messages: {
            forbidden: "Ignoring the buffer size is forbidden.",
        },
        schema: [],
        type: "problem",
    },
    name: "no-ignored-replay-buffer",
    create: (context) => {
        function checkNode(node, { arguments: args }) {
            if (!args || args.length === 0) {
                context.report({
                    messageId: "forbidden",
                    node,
                });
            }
        }
        return {
            "NewExpression > Identifier[name='ReplaySubject']": (node) => {
                const newExpression = eslint_etc_1.getParent(node);
                checkNode(node, newExpression);
            },
            "NewExpression > MemberExpression > Identifier[name='ReplaySubject']": (node) => {
                const memberExpression = eslint_etc_1.getParent(node);
                const newExpression = eslint_etc_1.getParent(memberExpression);
                checkNode(node, newExpression);
            },
            "CallExpression > Identifier[name=/^(publishReplay|shareReplay)$/]": (node) => {
                const callExpression = eslint_etc_1.getParent(node);
                checkNode(node, callExpression);
            },
        };
    },
});
module.exports = rule;
