// eslint-plugin-rxjs/dist/rules/no-nested-subscribe.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        docs: {
            category: "Best Practices",
            description: "Forbids the calling of `subscribe` within a `subscribe` callback.",
            recommended: "error",
        },
        fixable: undefined,
        messages: {
            forbidden: "Nested subscribe calls are forbidden.",
        },
        schema: [],
        type: "problem",
    },
    name: "no-nested-subscribe",
    create: (context) => {
        const { couldBeObservable } = eslint_etc_1.getTypeServices(context);
        const subscribeCallMap = new WeakMap();
        return {
            [`CallExpression > MemberExpression[property.name='subscribe']`]: (node) => {
                if (!couldBeObservable(node.object)) {
                    return;
                }
                const callExpression = eslint_etc_1.getParent(node);
                let parent = eslint_etc_1.getParent(callExpression);
                while (parent) {
                    if (subscribeCallMap.has(parent)) {
                        context.report({
                            messageId: "forbidden",
                            node: node.property,
                        });
                        return;
                    }
                    parent = eslint_etc_1.getParent(parent);
                }
                subscribeCallMap.set(callExpression);
            },
        };
    },
});
module.exports = rule;
