// eslint-plugin-rxjs/dist/rules/no-tap.js
"use strict";
const utils_1 = require("../utils");
const rule = utils_1.ruleCreator({
    defaultOptions: [],
    meta: {
        deprecated: true,
        docs: {
            category: "Best Practices",
            description: "Forbids the use of the `tap` operator.",
            recommended: false,
        },
        fixable: undefined,
        messages: {
            forbidden: "The tap operator is forbidden.",
        },
        replacedBy: ["ban-operators"],
        schema: [],
        type: "problem",
    },
    name: "no-tap",
    create: (context) => {
        return {
            "ImportDeclaration[source.value='rxjs/operators'] > ImportSpecifier[imported.name='tap']": (node) => {
                const { loc } = node;
                context.report({
                    messageId: "forbidden",
                    loc: {
                        ...loc,
                        end: {
                            ...loc.start,
                            column: loc.start.column + 3,
                        },
                    },
                });
            },
        };
    },
});
module.exports = rule;
