// eslint-plugin-rxjs/dist/rules/prefer-observer.js
"use strict";
const eslint_etc_1 = require("eslint-etc");
const utils_1 = require("../utils");
const defaultOptions = [];
const rule = utils_1.ruleCreator({
    defaultOptions,
    meta: {
        docs: {
            category: "Best Practices",
            description: "Forbids the passing separate callbacks to `subscribe` and `tap`.",
            recommended: false,
        },
        fixable: undefined,
        messages: {
            forbidden: "Passing separate callbacks is forbidden; pass an observer instead.",
        },
        schema: [
            {
                properties: {
                    allowNext: { type: "boolean" },
                },
                type: "object",
            },
        ],
        type: "problem",
    },
    name: "prefer-observer",
    create: (context, unused) => {
        const { couldBeFunction } = eslint_etc_1.getTypeServices(context);
        const [config = {}] = context.options;
        const { allowNext = true } = config;
        function checkArgs(callExpression, reportNode) {
            const { arguments: args } = callExpression;
            if (args.length > 1) {
                context.report({
                    messageId: "forbidden",
                    node: reportNode,
                });
            }
            else if (args.length === 1 && !allowNext) {
                const [arg] = args;
                if (eslint_etc_1.isArrowFunctionExpression(arg) ||
                    eslint_etc_1.isFunctionExpression(arg) ||
                    couldBeFunction(arg)) {
                    context.report({
                        messageId: "forbidden",
                        node: reportNode,
                    });
                }
            }
        }
        return {
            "CallExpression[callee.property.name='pipe'] > CallExpression[callee.name='tap']": (node) => checkArgs(node, node.callee),
            "CallExpression[callee.property.name='subscribe']": (node) => checkArgs(node, node.callee.property),
        };
    },
});
module.exports = rule;
